Zukini/EnumerableExtensions.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NUnit.Framework;
+
+namespace Zukini
+{
+    public static class EnumerableExtensions
+    {
+        /// <summary>
+        /// Shuffles items in Enumerable.
+        /// </summary>
+        /// <typeparam name="TSource"></typeparam>
+        /// <param name="source"></param>
+        /// <returns>IEnumerable</returns>
+        public static IEnumerable<TSource> Shuffle<TSource>(this IEnumerable<TSource> source)
+        {
+            return source.OrderBy(i => Guid.NewGuid());
+        }
+
+        /// <summary>
+        /// Retrieves random first item or throws error if item is not present.
+        /// </summary>
+        /// <typeparam name="TSource"></typeparam>
+        /// <param name="source"></param>
+        /// <param name="errorMessage"></param>
+        /// <returns>TSource</returns>
+        public static TSource RandomFirst<TSource>(this IEnumerable<TSource> source,
+            string errorMessage = "Cannot get random item: source is empty")
+        {
+            return source.Shuffle().FirstOrError(errorMessage);
+        }
+
+        /// <summary>
+        /// Returns first item or error if item is not present.
+        /// </summary>
+        /// <typeparam name="TSource"></typeparam>
+        /// <param name="source"></param>
+        /// <param name="errorMessage"></param>
+        /// <returns>TSource</returns>
+        public static TSource FirstOrError<TSource>(this IEnumerable<TSource> source, string errorMessage = "Item is not present.")
+        {
+            if (source == null) throw new ArgumentException("source");
+            var list = source as IList<TSource>;
+            if (list != null)
```

Comment:
IList<T> implements IEnumberable so this if block is not necessary - you should be able to simply execute the below else logic and have it work for both right?

Reply:
To be honest with you - I haven't created a wheel and modified FirstOrDefault:
https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs 

I think this is optimization. Seems like .Count is cheaper than creating whole Enumerator. I would keep it in line with MS approach. What me to remove this?